onnxruntime/core/session/inference_session.cc

```diff
@@ -553,7 +553,41 @@ int InferenceSession::GetCurrentNumRuns() const {
   return current_num_runs_.load();
 }
 
-common::Status InferenceSession::CheckTypes(MLDataType actual, MLDataType expected) {
+common::Status InferenceSession::CheckShapes(const std::string& input_name,
+                                             const TensorShape& input_shape,
+                                             const TensorShape& expected_shape) const {
+  auto input_shape_sz = input_shape.NumDimensions();
+  auto expected_shape_sz = expected_shape.NumDimensions();
+  if (input_shape_sz != expected_shape_sz) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Invalid rank for input: ",
+                           input_name,
+                           " Got: ", input_shape_sz, " Expected: ", expected_shape_sz);
+  }
+
+  std::vector<int> invalid_dim_indices;
+  for (size_t i = 0; i < input_shape_sz; ++i) {
+    if (expected_shape[i] < 0) {
+      continue;  // this represents a symbolic shape dimension
+    }
+    if (input_shape[i] != expected_shape[i]) {
+      invalid_dim_indices.push_back(i);
+    }
+  }
+
+  if (!invalid_dim_indices.empty()) {
+    std::ostringstream ostr;
+    ostr << "Got invalid dimensions for input: " << input_name << " for the following indices\n";
+    for (int i = 0, end = invalid_dim_indices.size(); i < end; ++i) {
```

Comment:
Wrong data type for i and end?  it should be size_t.

Reply:
fixed